orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java

```diff
@@ -0,0 +1,412 @@
+    orcType.setAttribute(ICEBERG_ID_ATTRIBUTE, String.valueOf(fieldId));
+    orcType.setAttribute(ICEBERG_REQUIRED_ATTRIBUTE, String.valueOf(isRequired));
+    return orcType;
+  }
+
+  /**
+   * Converts an Iceberg schema to a corresponding ORC schema within the context of an existing
+   * ORC file schema. This method also handles schema evolution from the original ORC file schema
+   * to the given Iceberg schema.
+   *
+   * @param schema an Iceberg schema
+   * @param originalOrcSchema an existing ORC file schema
+   * @return the resulting ORC schema
+   */
+  static TypeDescription toOrc(Schema schema, TypeDescription originalOrcSchema) {
+    final Map<Integer, OrcColumn> icebergToOrc = icebergToOrcMapping("root", originalOrcSchema);
+    return toOrc(Integer.MIN_VALUE, schema.asStruct(), true, icebergToOrc);
+  }
+
+  private static TypeDescription toOrc(Integer fieldId, Type type, boolean isRequired,
+                                       Map<Integer, OrcColumn> mapping) {
+    final TypeDescription orcType;
+
+    switch (type.typeId()) {
+      case STRUCT: {
+        orcType = TypeDescription.createStruct();
+        for (Types.NestedField nestedField : type.asStructType().fields()) {
+          String name = Optional.ofNullable(mapping.get(nestedField.fieldId()))
+              .map(OrcColumn::name)
+              .orElse(nestedField.name());
+          TypeDescription childType = toOrc(nestedField.fieldId(), nestedField.type(),
+              nestedField.isRequired(), mapping);
+          orcType.addField(name, childType);
+        }
+        break;
+      }
+      case LIST: {
+        Types.ListType list = (Types.ListType) type;
+        TypeDescription elementType = toOrc(list.elementId(), list.elementType(),
+            list.isElementRequired(), mapping);
+        orcType = TypeDescription.createList(elementType);
+        break;
+      }
+      case MAP: {
+        Types.MapType map = (Types.MapType) type;
+        TypeDescription keyType = toOrc(map.keyId(), map.keyType(), true, mapping);
+        TypeDescription valueType = toOrc(map.valueId(), map.valueType(), map.isValueRequired(),
+            mapping);
+        orcType = TypeDescription.createMap(keyType, valueType);
+        break;
+      }
+      default: {
+        if (mapping.containsKey(fieldId)) {
+          TypeDescription originalType = mapping.get(fieldId).type();
+          Optional<TypeDescription> promotedType = getPromotedType(type, originalType);
+
+          if (promotedType.isPresent()) {
+            orcType = promotedType.get();
+          } else {
+            Preconditions.checkArgument(isSameType(originalType, type),
+                "Can not promote %s type to %s",
+                originalType.getCategory(), type.typeId().name());
+            orcType = originalType.clone();
+          }
+        } else {
+          // This column is new, we need to convert via the usual way
+          orcType = toOrc(fieldId, type, isRequired);
```

Comment:
This should fail if `isRequired` is true because a required column is not present in the file schema.